File: src/lib/domain/csv.ts
import type { AdminRecord } from "./types.ts";

/** Quotes every cell and neutralises cells a spreadsheet would run as a formula. */
export function csvEscape(raw: string): string {
  let cell = raw;
  if (cell.length > 0 && "=+-@\t\r".includes(cell[0])) cell = "'" + cell;
  return '"' + cell.replaceAll('"', '""') + '"';
}

/** CSV of the rows on screen (search and filter included), as in the old panel. */
export function recordsToCsv(records: AdminRecord[], t: (key: string) => string = (k) => k): string {
  const header = ["ID", "Title", "Details", "Status", "Value", "Created"].map(t);
  const rows = records.map((r) => [
    r.id,
    r.title,
    r.subtitle,
    r.status ? t(r.status.key) : "",
    r.trailing ?? "",
    new Date(r.createdAt).toISOString(),
  ]);
  return [header, ...rows].map((row) => row.map(csvEscape).join(",")).join("\r\n") + "\r\n";
}

File: Packages/AdminKit/Sources/AdminDomain/CSV.swift
import Foundation

/// CSV export of the rows on screen. Ported from the old panel, including its
/// guard against spreadsheet formula injection.
public enum CSV {
    public static func make(_ records: [AdminRecord], translate: (String) -> String = { $0 }) -> String {
        var lines = [row([translate("ID"), translate("Title"), translate("Details"), translate("Status"), translate("Value"), translate("Created")])]
        let iso = ISO8601DateFormatter()
        for record in records {
            lines.append(row([
                record.id,
                record.title,
                record.subtitle,
                record.status.map { translate($0.key) } ?? "",
                record.trailing ?? "",
                iso.string(from: record.createdAt),
            ]))
        }
        return lines.joined(separator: "\r\n") + "\r\n"
    }

    static func row(_ cells: [String]) -> String {
        cells.map(escape).joined(separator: ",")
    }

    /// Quotes every cell and neutralises cells a spreadsheet would run as a formula.
    public static func escape(_ raw: String) -> String {
        var cell = raw
        if let first = cell.first, "=+-@\t\r".contains(first) {
            cell = "'" + cell
        }
        return "\"" + cell.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }
}
